### ISSUE TITLE
issue(s) with `center` argument of `rowVars` & `colVars`

### ISSUE THREAD

**Issue by jan-glx** [HenrikBengtsson/matrixStats#160]
Is `rowVars(mat-center)` supposed to be equal to `rowVars(mat, center)`?  Or only if `center==rowMeans(mat)`?

The former is not the case:
``` r
mat <- matrix(1:6, 2, 3)
matrixStats::rowVars(mat, center=1:2)
#> [1] 16 22
matrixStats::rowVars(mat-(1:2))
#> [1] 4 4
```

In any case, the current algorithm (with center) is numerically instable (@LTLA made me aware of that in a related case):
``` r
mat <- matrix(rnorm(100), 25, 4)+1E9
matrixStats::colVars(mat, center=colMeans(mat))
#> [1] 133.3333   0.0000   0.0000 133.3333
matrixStats::colVars(mat)
#> [1] 1.1828330 1.6575854 0.8001706 1.0915187
```

Furthermore, `center` accepts vectors of any length for which `max(length(center), nrow(mat))` is a multiple of `min(length(center), nrow(mat)`, leading to strange results:
``` r
mat <- matrix(1:6, 2, 3)
matrixStats::rowVars(mat, center=1:8)
#> [1]  16  22   4   4 -20 -26 -56 -68
````

`colVars` is equally affected.

I'll open a PR to highlight the issue in the code, but would only have time on the weekend to provide a full PR.

**Comment by jan-glx** [comment]
So far, `MatrixStats::rowVars(x, center=c)` has been calculating `[x^2] - c^2`* which is the same as `[(x-c)^2]` in case `c` is `0` or `[x]`:
```
[(x-c)^2] = [x^2 - 2 x c + c^2] = 
[x^2] - 2 [x] c + c^2 = [x^2] - c^2 <=>
- 2 [x] c + c^2 = - c^2 <=>
- 2 [x] c = - 2 c^2 <=>
[x] c = c^2 <=>
[x] = c | c=0
``` 
I think the latter is to be preferred: It would be consistent with [rowMads](https://github.com/HenrikBengtsson/matrixStats/blob/0bf5ebdcae7ca5908b5957d77b60c8a81771558f/R/rowMads.R#L25) and would allow to calculate mean squared error or variances with known mean (up to `N/(N-1)`). 
The only reasonable use cases of the current implementation that comes to my mind (`center=0` (to calculate sum of squares) or `center=mean(x)` ( to avoid re-computation of the mean for speed) would still yield identical results (apart from the (for the latter use-case) potentially severe numerical differences)).

I went ahead and updated the PR to a proper one: It changes the behavior of both `colVars` and `rowVars` with a center argument as described above. Existing tests run fine. I added some to test the new behavior and included it in the benchmark. The new implementation is (naturally) a bit slower, especially for small matrices. See [results](https://github.com/jan-glx/matrixStats/tree/benchmarkresults/reports).



*) Here, `[x]` denotes the arithmetic average of x (`[x]=1/N sum_{i=1}^N x_i`). 
To obtain unbiased estimates of the variances `E[(x-E[x])^2]`, `MatrixStats::rowVars` multiplies the computed average by `N/(N-1)`. This makes sense only if `center` is estimated as well, but that's a separate issue.

**Comment by HenrikBengtsson** [comment]
# Packages that use argument 'center'

I ran [revdepcheck on 307 reverse dependencies (150 from CRAN + 157 from BioConductor) w/ `MATRIXSTATS_CENTER_ONUSE=defunct`](https://github.com/HenrikBengtsson/matrixStats/blob/feature/arg-center/revdep/README.md).  The good news is that I found only four packages that use the `center` argument and, from manual code inspection, it looks like they're all using it correctly, i.e. they pass a correct estimate.

Conclusion: Changing the formula on how `rowVars(..., center)` calculates the variance will have no negative impact on CRAN and Bioconductor packages (at least for the ones that actually test their code).   It could of course be that there is non-tested code that is negatively impacted but that would require an amount of code inspection that I'm not willing to do.  What I'm still a bit concerned about is all the scripts out there that we can't check.  OTH, as I said before if they use `center` incorrectly, it's not that changing the formula will make it worse for them.


## antiProfiles (Bioconductor)

https://github.com/HCBravoLab/antiProfiles/blob/b9dbc2ad123b681824171efa011469b7e8a8029f/R/apStats.r#L55-L60

```r
    tMeds=sapply(tIndexes0[L0>=minL], function(ind) matrixStats::rowMedians(e, cols=ind))
    tMads=sapply(seq_along(tIndexes0[L0>=minL]), function(i) matrixStats::rowMads(e, cols=tIndexes0[L0>=minL][[i]], center=tMeds[,i]))
    colnames(tMads)=colnames(tMeds)

    meds0=matrixStats::rowMedians(tMeds)
    mads0=matrixStats::rowMedians(tMads)
```


## bigstep (CRAN)

https://github.com/pmszulc/bigstep/blob/8d39b578579526c83a968aa0b5fc122bcac1ef1c/R/reduce.R#L72-L73

```r
      means <- colMeans(XX)
      sds <- matrixStats::colSds(XX, center = means)
      XX <- t((t(XX) - means) / sds)
```


## expss (CRAN)

https://github.com/gdemin/expss/blob/7cec7fd56058daa4e0615aa8f42458f7a6ad97e2/R/weighted_stats.R#L109-L110

```r
w_mad = function(x, weight = NULL, na.rm = TRUE){
    internal_w_stat(x = x,  weight = weight, na.rm = na.rm, fun = function(x, w, na.rm){
        matrixStats::weightedMad(x = x, w = w, na.rm = na.rm,
                                 center = w_median(x = x, weight = w, na.rm = na.rm)
        )
    })
}
```

Comment: This use of `center` is unnecessary; if not specified, the exact same calculation is done by `matrixStats::weightedMad()` internally.  It's could also be that it is faster to leave it out if `na.rm=TRUE` since NAs are then removed ones instead of twice.


## Seurat (CRAN)

https://github.com/satijalab/seurat/blob/fe93b05745e55ec2f66e3f0b4c4196aad9f4d5a7/R/utilities.R#L928-L933

```r
FastRowScale <- function(
  mat,
  center = TRUE,
  scale = TRUE,
  scale_max = 10
) {
  # inspired by https://www.r-bloggers.com/a-faster-scale-function/
  if (center) {
    rm <- rowMeans2(x = mat, na.rm = TRUE)
  }
  if (scale) {
    if (center) {
      rsd <- rowSds(mat, center = rm)
```

**Comment by HenrikBengtsson** [comment]
To go ahead, I think I'll have `rowVars()` internally validate that the results from switching to the (new) "primary" formula is identical to when using the (current) "alternative" formula.  If not, an informative warning will be produced.  In order for performance not to suffer, this validation should only be done once per session, or occasionally, until after a long time has been established there are no reports.  

This behavior could be controlled via an option, e.g. `getOption("matrixStats.rowVars.center.validate", "once")` accepting also `"never"`, `"always"`, or fraction of times as given by a number in [0,1], where `0 == "never"` and `1 == "always"`.    The problem with `"once"` is that the usage elsewhere in the software stack might be masked by a single proper usage.

**Comment by HenrikBengtsson** [comment]
While working on this, I realized that the original rationale for using the alternative formula:

```
x <- x * x
x <- rowMeans(x, na.rm = na.rm)
x <- (x - center^2)
x * (n/(n - 1))
```

instead of the primary formula:

```
x <- X
x <- (x - center)^2
x <- rowMeans(x, na.rm = na.rm)
x * (n/(n - 1))
```

was probably that the former would only have to allocate one full matrix (`x <- x*x`), whereas the latter would allocate two full matrices (`(x - center)` and then `(x - center)^2`).

However, from some quick `profmem::profmem()` tests, it turns out that R is clever enough to know that it can reuse the existing allocation, i.e. there's only one full matrix allocation in both cases.  It could be that things worked differently in the past - for instance prior to the reference counting updates - don't know.

**Comment by HenrikBengtsson** [comment]
... another reason was probably that:

```r
x <- x * x
x <- {col,row}Means(x, na.rm = na.rm)
x <- (x - center^2)
x * (n/(n - 1))
```
works regardless of rows or columns, whereas `(x - center)` only works out of the box for `rowVars()`.  Special care needs to be taken for `colVars()`, especially if one wants to be conservative about memory allocations.

**Comment by HenrikBengtsson** [comment]
I've now incorporated this in the 'develop' branch (= next release);

SIGNIFICANT CHANGES:

 * colVars() and rowVars() with argument 'center' now calculates the sample
   variance using the n/(n-1)*avg((x-center)^2) formula rather than the
   n/(n-1)*(avg(x^2)-center^2) formula that was used in the past.  Both give
   the same result when 'center' is the correct sample mean estimate.  The
   main reason for this change is that, if an incorrect 'center' is provided,
   in contrast to the old approach, the new approach is guaranteed to give at
   least non-negative results, despite being incorrect.
   BACKWARD COMPATIBILITY: Out of all 307 reverse dependencies on CRAN and
   Bioconductor, only four called these functions with argument 'center'. All
   of them pass their package checks also after this update.  To further
   protect against a negative impact in existing user scripts, colVars() and
   rowVars() will calculate both versions and assert that the result is the
   same.  If not, an informative error is produced.  To limit the performance
   impact, this validation is run only once every 50:th call, a frequency that
   can be controlled by R option 'matrixStats.vars.formula.freq'.  Setting it
   to 0 or NULL will disable the validation.  The default can also be
   controlled by environment variable 'R_MATRIXSTATS_VARS_FORMULA_FREQ'.
   This validation framework will be removed in a future version of the package
   after it has been established that this change has no negative impact.

I've rerun revdep checks on 307 packages with `R_MATRIXSTATS_VARS_FORMULA_FREQ=1` and things worked out as expected, cf. https://github.com/HenrikBengtsson/matrixStats/blob/1b68d5cb946f2c8c86522b2e19837f99cbc7425d/revdep/README.md.  It reveals some thinkos in the MatrixGenerics package tests, which has been discussed elsewhere.

Thxs again for pushing this.
